faer-libs/faer-core/src/complex_native/c32_impl.rs

```diff
@@ -0,0 +1,1051 @@
+
+#[cfg(feature = "std")]
+macro_rules! impl_from_num_complex {
+    ($( $method:ident ( self $( , $arg:ident : $ty:ty )* ) -> $ret:ty ; )*) => {
+        $(
+            #[inline(always)]
+            pub fn $method(self $( , $arg : $ty )* ) -> $ret {
+                self.to_num_complex().$method( $( $arg , )* ).into()
+            }
+        )*
+    };
+}
+
+// 32-bit complex floating point type. See the module-level documentation for more details.
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone, PartialEq)]
+#[repr(C)]
+pub struct c32 {
+    pub re: f32,
+    pub im: f32,
+}
+
+impl c32 {
+    /// Create a new complex number.
+    #[inline(always)]
+    pub fn new(re: f32, im: f32) -> Self {
+        Self { re, im }
+    }
+
+    /// Construct the imaginary number.
+    #[inline(always)]
+    pub fn i() -> Self {
+        Self::new(0.0, 1.0)
+    }
+
+    /// Create a complex number from a phase.
+    #[cfg(feature = "std")]
+    #[inline(always)]
+    pub fn cis(phase: f32) -> Self {
+        Self::new(phase.cos(), phase.sin())
+    }
+
+    /// Create a complex number from polar coordinates.
+    #[cfg(feature = "std")]
+    #[inline(always)]
+    pub fn from_polar(r: f32, theta: f32) -> Self {
+        Self::new(r * theta.cos(), r * theta.sin())
+    }
+
+    /// Convert the number to a num_complex::Complex32.
+    #[inline(always)]
+    pub fn to_num_complex(self) -> num_complex::Complex<f32> {
+        self.into()
+    }
+
+    /// Returns the real part of the complex number.
+    #[inline(always)]
+    pub fn re(self) -> f32 {
+        self.re
+    }
+
+    /// Returns the imaginary part of the complex number.
+    #[inline(always)]
+    pub fn im(self) -> f32 {
+        self.im
+    }
+
+    /// Calculate the complex conjugate of self.
+    #[inline(always)]
+    pub fn conj(self) -> Self {
+        self.faer_conj()
+    }
+
+    #[cfg(feature = "std")]
+    impl_from_num_complex!(
+        is_nan(self) -> bool;
+        is_infinite(self) -> bool;
+        is_finite(self) -> bool;
+        is_normal(self) -> bool;
+        recip(self) -> Self;
+        powi(self, exp: i32) -> Self;
+        powu(self, exp: u32) -> Self;
+        powf(self, exp: f32) -> Self;
+        powc(self, exp: num_complex::Complex<f32>) -> Self;
+        sqrt(self) -> Self;
+        exp(self) -> Self;
+        exp2(self) -> Self;
+        expf(self, base: f32) -> Self;
+        ln(self) -> Self;
+        log(self, base: f32) -> Self;
+        log2(self) -> Self;
+        log10(self) -> Self;
+        cbrt(self) -> Self;
+        sin(self) -> Self;
+        cos(self) -> Self;
+        tan(self) -> Self;
+        asin(self) -> Self;
+        acos(self) -> Self;
+        atan(self) -> Self;
+        sinh(self) -> Self;
+        cosh(self) -> Self;
+        tanh(self) -> Self;
+        asinh(self) -> Self;
+        acosh(self) -> Self;
+        atanh(self) -> Self;
+        abs(self) -> f32;
+        arg(self) -> f32;
+        norm(self) -> f32;
```

Comment:
maybe forward to `faer_norm` instead for better numerical stability, since this is a commonly used function